torch_xla/csrc/runtime/util.h

```diff
@@ -128,6 +130,53 @@ T Multiply(const S& input) {
                          std::multiplies<T>());
 }
 
+namespace internal {
+
+// ExtractValue<U>::type is T if U is absl::StatusOr<T>, and is undefined
+// otherwise.
+template <typename U>
+struct ExtractValue;
+template <typename T>
+struct ExtractValue<absl::StatusOr<T>> {
+  using type = T;
+};
+
+}  // namespace internal
+
+// RaisePythonValueErrorOnFailure(func) requires `func` to be a functor that
+// takes no argument and returns an absl::StatusOr<T>. It's a wrapper of
+// `func()` that translates any failure in `func()` to a Python ValueError
+// exception. In particular:
+//
+//   - if `func()` returns an error, throws an std::invalid_argument,
+//     which is translated to a Python ValueError exception;
+//     (https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html).
+//   - if `func()` throws any exception, rethrows it as an
+//     std::invalid_argument so that we get a Python ValueError;
+//   - if `func()` successfully returns a value of type T, returns the value;
+//   - however, if `func()` crashes (e.g. due to a CHECK), we cannot
+//     catch it; therefore we should ensure that `func()` never
+//     crashes (and fix any crash as a bug).
+template <typename Func>
+typename internal::ExtractValue<decltype(std::declval<Func>()())>::type
+RaisePythonValueErrorOnFailure(const Func& func) {
+  decltype(std::declval<Func>()()) result;
+  try {
+    result = func();
```

Comment:
Can probably simply a bit by moving logics using result inside; so this can be

```
auto result = func();
 if (result.ok()) {
    return *std::move(result);
 }
 throw std::invalid_argument(std::string(result.status().message()));
```
and the decltype in 163 can be removed.

The invalid_argument will get caught in 166 and rethrow with the same message in 167.

Reply:
I considered this and actually think the re-throw makes the logic more complex (compared with the `decltype`) and slightly less efficient.